File: content/examples/other/leetcode/problems/set-matrix-zeroes/example-1/Solution.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class Solution {
  static int[][] setZeroesWithSets(int[][] matrix) {
    int rows = matrix.length;
    int cols = rows == 0 ? 0 : matrix[0].length;
    Set<Integer> zeroRows = new HashSet<>();
    Set<Integer> zeroCols = new HashSet<>();

    for (int r = 0; r < rows; r += 1) {
      for (int c = 0; c < cols; c += 1) {
        if (matrix[r][c] == 0) {
          zeroRows.add(r);
          zeroCols.add(c);
        }
      }
    }

    for (int r = 0; r < rows; r += 1) {
      for (int c = 0; c < cols; c += 1) {
        if (zeroRows.contains(r) || zeroCols.contains(c)) matrix[r][c] = 0;
      }
    }
    return matrix;
  }

  public static void main(String[] args) {
    int[][] out = setZeroesWithSets(new int[][] {{1, 1, 1}, {1, 0, 1}, {1, 1, 1}});
    System.out.println(Arrays.deepToString(out));
  }
}

File: content/examples/other/leetcode/problems/set-matrix-zeroes/example-1/solution.js
function setZeroesWithSets(matrix) {
  const rows = matrix.length;
  const cols = rows === 0 ? 0 : matrix[0].length;
  const zeroRows = new Set();
  const zeroCols = new Set();

  for (let r = 0; r < rows; r += 1) {
    for (let c = 0; c < cols; c += 1) {
      if (matrix[r][c] === 0) {
        zeroRows.add(r);
        zeroCols.add(c);
      }
    }
  }

  for (let r = 0; r < rows; r += 1) {
    for (let c = 0; c < cols; c += 1) {
      if (zeroRows.has(r) || zeroCols.has(c)) matrix[r][c] = 0;
    }
  }
  return matrix;
}

if (require.main === module) {
  console.log(setZeroesWithSets([[1, 1, 1], [1, 0, 1], [1, 1, 1]]));
}

module.exports = { setZeroesWithSets };
